In [139]:
import pandas as pd
import numpy as np
from datetime import datetime, timedelta

In [461]:
# 중국 23년 데이터 읽기
df2301 = pd.read_csv(f"C:/Users/ITSC/Desktop/ChrisProject/FinalProject/Git/데이터/중국기상데이터/중국 기상 202301.csv", sep=',',encoding='mbcs')
df2302 = pd.read_csv(f"C:/Users/ITSC/Desktop/ChrisProject/FinalProject/Git/데이터/중국기상데이터/중국 기상 202302.csv", sep=',',encoding='mbcs')
df2303 = pd.read_csv(f"C:/Users/ITSC/Desktop/ChrisProject/FinalProject/Git/데이터/중국기상데이터/중국 기상 202303.csv", sep=',',encoding='mbcs')
df2304 = pd.read_csv(f"C:/Users/ITSC/Desktop/ChrisProject/FinalProject/Git/데이터/중국기상데이터/중국 기상 202304.csv", sep=',',encoding='mbcs')
df2305 = pd.read_csv(f"C:/Users/ITSC/Desktop/ChrisProject/FinalProject/Git/데이터/중국기상데이터/중국 기상 202305.csv", sep=',',encoding='mbcs')
df2306 = pd.read_csv(f"C:/Users/ITSC/Desktop/ChrisProject/FinalProject/Git/데이터/중국기상데이터/중국 기상 202306.csv", sep=',',encoding='mbcs')
df2307 = pd.read_csv(f"C:/Users/ITSC/Desktop/ChrisProject/FinalProject/Git/데이터/중국기상데이터/중국 기상 202307.csv", sep=',',encoding='mbcs')
df2308 = pd.read_csv(f"C:/Users/ITSC/Desktop/ChrisProject/FinalProject/Git/데이터/중국기상데이터/중국 기상 202308.csv", sep=',',encoding='mbcs')
df2309 = pd.read_csv(f"C:/Users/ITSC/Desktop/ChrisProject/FinalProject/Git/데이터/중국기상데이터/중국 기상 202309.csv", sep=',',encoding='mbcs')
df2310 = pd.read_csv(f"C:/Users/ITSC/Desktop/ChrisProject/FinalProject/Git/데이터/중국기상데이터/중국 기상 202310.csv", sep=',',encoding='mbcs')
df2311 = pd.read_csv(f"C:/Users/ITSC/Desktop/ChrisProject/FinalProject/Git/데이터/중국기상데이터/중국 기상 202311.csv", sep=',',encoding='mbcs')
df2312 = pd.read_csv(f"C:/Users/ITSC/Desktop/ChrisProject/FinalProject/Git/데이터/중국기상데이터/중국 기상 202312.csv", sep=',',encoding='mbcs')

In [462]:
df2301.head()

지점     지점명                일시    습도  16 방위 풍향   풍속  강수량   현지기압    해면기압  \
0  50527  HAILAR  2023-01-01 00:00  76.3      29.0  3.0  NaN  949.0  1037.3   
1  50527  HAILAR  2023-01-01 03:00  69.1      25.0  5.0  NaN  950.4  1038.6   
2  50527  HAILAR  2023-01-01 06:00  64.7      25.0  6.0  NaN  950.0  1038.0   
3  50527  HAILAR  2023-01-01 09:00  70.8      25.0  7.0  NaN  950.0  1038.0   
4  50527  HAILAR  2023-01-01 12:00  71.1      28.0  4.0  NaN  950.6  1039.1   

   변화량 기압  이슬점 온도    기온  최고 기온  최저 기온  
0     1.4   -30.0 -27.1  -19.4  -27.4  
1     1.4   -27.2 -23.1  -19.4  -27.4  
2     0.4   -25.7 -20.8  -19.4  -27.4  
3     0.0   -25.1 -21.2  -20.3  -27.4  
4     0.6   -26.5 -22.7  -20.3  -27.4

- 남길 피처들의 목록 : 지점, 일시(월), 일시(시간), 풍향, 풍속, 강수량, 기온

In [463]:
# 지점 목록 확인
df2301["지점명"].unique()

array(['HAILAR', '치치하르', 'TONGHE', '알타이', 'KARAMAY', '이닝', '우루무치', 'KUQA',
       'KORLA', 'KASHI', 'TAZHONG', 'RUOQIANG', 'HOTAN', 'HAMI',
       'EJIN QI', 'MAZONG SHAN', '둔황', 'BAYAN MOD', 'JIUQUAN', 'MINQIN',
       'DULAN', '시닝', 'YU ZHONG', 'ERENHOT', 'HALIUT', 'HOHHOT',
       'YINCHUAN', '타이위안치', 'YAN AN', 'JARUD QI', 'XILIN HOT', '창춘', '치펭',
       '연길', '센양', '베이징', '텐진', '달리안', '지난', '칭따오', 'SHIQUANHE', 'XAINZA',
       '라싸', 'TUOTUOHE', 'YUSHU', 'DARLAG', "RUO'ERGAI", 'SOG XIAN',
       'QAMDO', 'WENJIANG', 'XICHANG', 'TENGCHONG', '쿤밍', 'SIMAO',
       'MENGZI', 'ZHENGZHOU', '한종', 'JINGHE', '이창', '우한', '충칭', '장사',
       'ZHIJIANG', '귀양', 'GANZHOU', '쑤조우', '난징', '상하이', '난창', 'QU XIAN',
       'DACHEN DAO', '푸조우', '타이베이', 'BAISE', '광조우', 'SHANTOU', 'NANNING',
       '하이코우', 'YAXIAN', 'XISHA DAO'], dtype=object)

- 포함할 도시 목록
    - 베이징, 지난, 칭따오, 쑤조우, 난징, 상하이, DACHEN DAO, 푸조우, 광조우 

In [26]:
# 데이터 concat
df = pd.concat([df2301, df2302, df2303, df2304, df2305, df2306, df2307, df2308, df2309, df2310, df2311, df2312], axis = 0)

### 데이터 확인

In [27]:
df.head()

지점     지점명                일시    습도  16 방위 풍향   풍속  강수량   현지기압    해면기압  \
0  50527  HAILAR  2023-01-01 00:00  76.3      29.0  3.0  NaN  949.0  1037.3   
1  50527  HAILAR  2023-01-01 03:00  69.1      25.0  5.0  NaN  950.4  1038.6   
2  50527  HAILAR  2023-01-01 06:00  64.7      25.0  6.0  NaN  950.0  1038.0   
3  50527  HAILAR  2023-01-01 09:00  70.8      25.0  7.0  NaN  950.0  1038.0   
4  50527  HAILAR  2023-01-01 12:00  71.1      28.0  4.0  NaN  950.6  1039.1   

   변화량 기압  이슬점 온도    기온  최고 기온  최저 기온  
0     1.4   -30.0 -27.1  -19.4  -27.4  
1     1.4   -27.2 -23.1  -19.4  -27.4  
2     0.4   -25.7 -20.8  -19.4  -27.4  
3     0.0   -25.1 -21.2  -20.3  -27.4  
4     0.6   -26.5 -22.7  -20.3  -27.4

In [29]:
df.shape

(187390, 14)

In [35]:
df.describe()

지점             습도       16 방위 풍향             풍속  \
count  187390.000000  183988.000000  187325.000000  187318.000000   
mean    55305.846176      58.361687      18.263857       2.447949   
std      2706.845439      25.715517      11.455992       4.994962   
min     50527.000000     -99.800000    -998.000000    -998.000000   
25%     52866.000000      38.000000       9.000000       1.000000   
50%     55228.000000      60.900000      18.000000       2.000000   
75%     57687.000000      79.900000      28.000000       3.000000   
max     59981.000000      99.400000      36.000000      22.000000   

                강수량           현지기압           해면기압         변화량 기압  \
count  44610.000000  187328.000000  187296.000000  187311.000000   
mean    -269.107608     909.885057    1015.691807       1.045499   
std      447.420921     114.392753      14.136381       1.010279   
min     -999.000000     -99.800000     -99.800000     -99.800000   
25%     -999.000000     864.900000    1006.400000       0.400000   
50%        0.400000     943.000000    1014.700000       0.900000   
75%        3.000000     999.500000    1024.000000       1.500000   
max      355.000000    1044.600000    1074.400000      10.100000   

              이슬점 온도             기온          최고 기온          최저 기온  
count  187328.000000  187327.000000  185226.000000  185225.000000  
mean        3.513499      12.751792      18.405924       7.353979  
std        14.788932      13.357964      13.026095      13.945975  
min       -99.800000     -99.800000     -99.800000     -99.800000  
25%        -7.900000       3.800000       9.800000      -1.900000  
50%         5.200000      15.000000      21.100000       9.500000  
75%        15.900000      23.300000      28.375000      18.400000  
max        29.600000      42.900000      43.800000      31.800000

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 187390 entries, 0 to 17234
Data columns (total 14 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   지점        187390 non-null  int64  
 1   지점명       187390 non-null  object 
 2   일시        187390 non-null  object 
 3   습도        183988 non-null  float64
 4   16 방위 풍향  187325 non-null  float64
 5   풍속        187318 non-null  float64
 6   강수량       44610 non-null   float64
 7   현지기압      187328 non-null  float64
 8   해면기압      187296 non-null  float64
 9   변화량 기압    187311 non-null  float64
 10  이슬점 온도    187328 non-null  float64
 11  기온        187327 non-null  float64
 12  최고 기온     185226 non-null  float64
 13  최저 기온     185225 non-null  float64
dtypes: float64(11), int64(1), object(2)
memory usage: 21.4+ MB


In [233]:
# 사용하기로 한 관측소 목록
loc_list = ["베이징", "지난", "칭따오", "쑤조우", "난징", "상하이", "DACHEN DAO", "푸조우", "광조우"]

In [234]:
# 관측소 목록으로 1차 필터링
ndf = df[df["지점명"].isin(loc_list)].reset_index(drop = True)

In [235]:
ndf.head()

지점  지점명                일시    습도  16 방위 풍향   풍속  강수량    현지기압    해면기압  \
0  54511  베이징  2023-01-01 00:00  51.9       6.0  1.0  NaN  1032.0  1036.5   
1  54511  베이징  2023-01-01 03:00  38.1       4.0  3.0  NaN  1032.8  1037.2   
2  54511  베이징  2023-01-01 06:00  31.1       6.0  3.0  NaN  1030.4  1034.8   
3  54511  베이징  2023-01-01 09:00  34.9      22.0  1.0  NaN  1031.3  1035.7   
4  54511  베이징  2023-01-01 12:00  54.1      36.0  0.0  NaN  1033.2  1037.7   

   변화량 기압  이슬점 온도   기온  최고 기온  최저 기온  
0     1.6   -14.2 -5.9    7.7   -9.1  
1     0.8   -11.5  1.2    7.7   -8.0  
2     2.4   -12.1  3.4    7.7   -8.0  
3     0.9   -12.6  1.2    4.6   -8.0  
4     1.9   -12.2 -4.3    4.6   -8.0

In [236]:
# 프로토타입용 피처만 남기기
ndf = ndf.loc[:, ["지점", "일시", "16 방위 풍향", "풍속", "강수량", "기온"]]

In [237]:
# 월, 시간 피처 추가
ndf["month"] = ndf["일시"].str.slice(start = 5, stop = 7)
ndf["time"] = ndf["일시"].str.slice(start = 11, stop = 13)

In [238]:
ndf.head()

지점                일시  16 방위 풍향   풍속  강수량   기온 month time
0  54511  2023-01-01 00:00       6.0  1.0  NaN -5.9    01   00
1  54511  2023-01-01 03:00       4.0  3.0  NaN  1.2    01   03
2  54511  2023-01-01 06:00       6.0  3.0  NaN  3.4    01   06
3  54511  2023-01-01 09:00      22.0  1.0  NaN  1.2    01   09
4  54511  2023-01-01 12:00      36.0  0.0  NaN -4.3    01   12

In [239]:
# 강수량 결측치 0 값으로 대체
ndf["강수량"] = ndf["강수량"].fillna(value = 0)

In [240]:
ndf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22360 entries, 0 to 22359
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   지점        22360 non-null  int64  
 1   일시        22360 non-null  object 
 2   16 방위 풍향  22359 non-null  float64
 3   풍속        22359 non-null  float64
 4   강수량       22360 non-null  float64
 5   기온        22359 non-null  float64
 6   month     22360 non-null  object 
 7   time      22360 non-null  object 
dtypes: float64(4), int64(1), object(3)
memory usage: 1.4+ MB


- 26280개의 데이터가 있어야 하지만 22360개로 약 4000개의 데이터가 결측 상태
    - 365 * 8 * 9 = 26,280
    
- 22360개의 데이터 중에도 결측치가 있음
    - 풍향 1개의 결측치
    - 풍속 1개의 결측치
        - 풍향과 풍속은 값이 0인 경우도 추가 확인이 필요함
    - 기온 1개의 결측치

In [241]:
# 풍향 결측치 데이터 확인
ndf[ndf["16 방위 풍향"].isna()]

지점                일시  16 방위 풍향  풍속  강수량  기온 month time
5831  58238  2023-04-14 03:00       NaN NaN  0.0 NaN    04   03

- 3가지 항목이 결측치인 경우가 다행히 동일한 데이터
- 또한 3개의 데이터가 서로 시간 혹은 지점이 다르기에 처리하기 약간 용이할 것으로 보임
- 각 데이터의 전후 2개 데이터를 구한 후 평균값으로 처리할 예정

In [242]:
ndf[5829:5834]

지점                일시  16 방위 풍향   풍속    강수량    기온 month time
5829  58238  2023-04-06 15:00      29.0  1.0    0.5  10.5    04   15
5830  58238  2023-04-06 18:00      27.0  1.0    0.4   9.6    04   18
5831  58238  2023-04-14 03:00       NaN  NaN    0.0   NaN    04   03
5832  58238  2023-04-14 06:00       1.0  4.0 -999.0  21.1    04   06
5833  58238  2023-04-14 09:00      33.0  3.0    0.0  22.2    04   09

In [243]:
ndf.loc[5831, "16 방위 풍향"] = 35.0
ndf.loc[5831, "풍속"] = 3.0
ndf.loc[5831, "기온"] = 20.0

In [244]:
ndf[5829:5834]

지점                일시  16 방위 풍향   풍속    강수량    기온 month time
5829  58238  2023-04-06 15:00      29.0  1.0    0.5  10.5    04   15
5830  58238  2023-04-06 18:00      27.0  1.0    0.4   9.6    04   18
5831  58238  2023-04-14 03:00      35.0  3.0    0.0  20.0    04   03
5832  58238  2023-04-14 06:00       1.0  4.0 -999.0  21.1    04   06
5833  58238  2023-04-14 09:00      33.0  3.0    0.0  22.2    04   09

### 이상치 확인해보기

In [245]:
ndf.describe()

지점      16 방위 풍향            풍속           강수량            기온
count  22360.000000  22360.000000  22360.000000  22360.000000  22360.000000
mean   57287.061360     18.057782      2.698166    -62.825894     17.654763
std     1846.779921     11.238717      2.162212    245.864336     10.249197
min    54511.000000      0.000000      0.000000   -999.000000    -15.600000
25%    54857.000000      9.000000      1.000000      0.000000      9.900000
50%    58238.000000     17.000000      2.000000      0.000000     19.600000
75%    58666.000000     29.000000      3.000000      0.000000     26.000000
max    59287.000000     36.000000     22.000000    355.000000     40.500000

- 현재 10개 지점의 데이터에서는 풍향, 풍속, 기온은 이상치가 없는 것으로 확인됨
- 강수량의 경우 최솟값이 -999로 상식적으로 발생할 수 없는 값이 입력되어 있음
    - 강수량 값이 음수일 경우 0으로 대체할 예정

In [246]:
ndf["강수량"] = ndf["강수량"].map(lambda x: 0 if x < 0 else x)

In [247]:
ndf.describe()

지점      16 방위 풍향            풍속           강수량            기온
count  22360.000000  22360.000000  22360.000000  22360.000000  22360.000000
mean   57287.061360     18.057782      2.698166      1.555098     17.654763
std     1846.779921     11.238717      2.162212      8.639113     10.249197
min    54511.000000      0.000000      0.000000      0.000000    -15.600000
25%    54857.000000      9.000000      1.000000      0.000000      9.900000
50%    58238.000000     17.000000      2.000000      0.000000     19.600000
75%    58666.000000     29.000000      3.000000      0.000000     26.000000
max    59287.000000     36.000000     22.000000    355.000000     40.500000

In [248]:
# 강수량의 값이 0이 아닌 경우의 강수량 describe 확인해보기
ndf[ndf["강수량"] != 0].describe()

지점     16 방위 풍향           풍속          강수량           기온
count   4292.000000  4292.000000  4292.000000  4292.000000  4292.000000
mean   57849.269105    18.159832     2.891193     8.101584    19.581733
std     1617.337206    12.137125     2.452541    18.326089     8.308238
min    54511.000000     0.000000     0.000000     0.100000   -13.100000
25%    58027.000000     7.000000     1.000000     0.400000    14.600000
50%    58362.000000    16.000000     2.000000     2.000000    21.700000
75%    58847.000000    30.000000     3.000000     8.000000    25.900000
max    59287.000000    36.000000    22.000000   355.000000    37.800000

- 강수량 355는 비현실적으로 높은 수치인것으로 생각됨 추가 확인이 필요함

In [249]:
ndf[ndf["강수량"] > 100]

지점                일시  16 방위 풍향   풍속    강수량    기온 month time
11313  58238  2023-07-17 00:00       6.0  2.0  104.0  23.4    07   00
11314  58238  2023-07-17 03:00       1.0  2.0  118.0  23.4    07   03
11315  58238  2023-07-17 06:00       7.0  2.0  149.0  25.2    07   06
12146  58847  2023-07-28 15:00      12.0  4.0  145.0  24.6    07   15
12147  58847  2023-07-28 18:00      32.0  1.0  225.0  24.9    07   18
12148  58847  2023-07-28 21:00      35.0  2.0  277.0  24.8    07   21
12149  58847  2023-07-29 00:00      12.0  2.0  275.0  24.0    07   00
12150  58847  2023-07-29 03:00      11.0  3.0  197.0  24.7    07   03
12151  58847  2023-07-29 06:00      16.0  2.0  111.0  26.8    07   06
13482  58238  2023-08-22 03:00       2.0  1.0  114.0  24.4    08   03
15709  58847  2023-09-05 18:00      30.0  3.0  324.0  23.0    09   18
15710  58847  2023-09-05 21:00      29.0  2.0  355.0  22.9    09   21
15711  58847  2023-09-06 00:00      36.0  1.0  352.0  23.3    09   00
15712  58847  2023-09-06 03:00      25.0  2.0  193.0  23.9    09   03
15905  59287  2023-09-07 18:00      10.0  3.0  131.0  24.6    09   18
15906  59287  2023-09-07 21:00      10.0  2.0  145.0  24.4    09   21
15907  59287  2023-09-08 00:00       8.0  2.0  144.0  24.4    09   00

In [250]:
df[df["지점"] == 58847].head()

지점  지점명                일시    습도  16 방위 풍향   풍속  강수량    현지기압    해면기압  \
17349  58847  푸조우  2023-01-01 00:00  65.8      36.0  1.0  NaN  1018.8  1029.2   
17350  58847  푸조우  2023-01-01 03:00  66.1      33.0  1.0  NaN  1018.8  1029.2   
17351  58847  푸조우  2023-01-01 06:00  66.1      35.0  3.0  NaN  1016.2  1026.6   
17352  58847  푸조우  2023-01-01 09:00  66.1       5.0  1.0  NaN  1016.4  1026.8   
17353  58847  푸조우  2023-01-01 12:00  67.0      29.0  1.0  NaN  1018.0  1028.4   

       변화량 기압  이슬점 온도    기온  최고 기온  최저 기온  
17349     1.2     6.2  12.4   14.4   10.2  
17350     0.0     7.2  13.4   14.4   11.7  
17351     2.6     7.5  13.7   14.2   11.7  
17352     0.2     7.3  13.5   13.9   11.7  
17353     1.6     7.4  13.4   13.9   11.7

- 300이상의 강수량을 보이는 지역은 푸조우(푸저우)인 것으로 확인되며 당시 9월 초에 태풍 사올라가 지나간 기사를 확인할 수 있음
    - 기사 링크 https://www.jejuilbo.net/news/articleView.html?idxno=212704

### 구간이 결측치인 경우 확인해보기

In [251]:
ndf.head()

지점                일시  16 방위 풍향   풍속  강수량   기온 month time
0  54511  2023-01-01 00:00       6.0  1.0  0.0 -5.9    01   00
1  54511  2023-01-01 03:00       4.0  3.0  0.0  1.2    01   03
2  54511  2023-01-01 06:00       6.0  3.0  0.0  3.4    01   06
3  54511  2023-01-01 09:00      22.0  1.0  0.0  1.2    01   09
4  54511  2023-01-01 12:00      36.0  0.0  0.0 -4.3    01   12

In [252]:
# 각 지점별로 결측치의 수량 확인해보기
ndf.groupby("지점").count()

일시  16 방위 풍향    풍속   강수량    기온  month  time
지점                                                  
54511  2498      2498  2498  2498  2498   2498  2498
54823  2499      2499  2499  2499  2499   2499  2499
54857  2482      2482  2482  2482  2482   2482  2482
58027  2461      2461  2461  2461  2461   2461  2461
58238  2486      2486  2486  2486  2486   2486  2486
58362  2490      2490  2490  2490  2490   2490  2490
58666  2460      2460  2460  2460  2460   2460  2460
58847  2495      2495  2495  2495  2495   2495  2495
59287  2489      2489  2489  2489  2489   2489  2489

- 대체로 2400개 후반대의 데이터가 관측되었음

In [254]:
# 일시 피처의 데이터 타입을 datetime으로 변환
ndf["일시"] = ndf["일시"].map(lambda x: datetime.strptime(x, '%Y-%m-%d %H:%M'))

In [255]:
time_list = [[ndf.loc[0, "일시"] + timedelta(hours = 3) * i] for i in range(365*8)]

In [256]:
for i in range(len(time_list)):
    time_list[i].append(len(ndf[ndf["일시"] == time_list[i][0]]))

In [257]:
time_df = pd.DataFrame(time_list)

In [263]:
time_df.head()

0  1
0 2023-01-01 00:00:00  9
1 2023-01-01 03:00:00  9
2 2023-01-01 06:00:00  9
3 2023-01-01 09:00:00  9
4 2023-01-01 12:00:00  9

In [260]:
time_df[1].map(lambda x: abs(9 - x)).sum()

3920

In [264]:
time_df[time_df[1] != 9]

0  1
33   2023-01-05 03:00:00  3
138  2023-01-18 06:00:00  3
157  2023-01-20 15:00:00  3
327  2023-02-10 21:00:00  8
334  2023-02-11 18:00:00  2
...                  ... ..
2825 2023-12-20 03:00:00  7
2877 2023-12-26 15:00:00  7
2882 2023-12-27 06:00:00  1
2886 2023-12-27 18:00:00  5
2901 2023-12-29 15:00:00  8

[515 rows x 2 columns]

In [269]:
time_df.columns = ["datetime", "count"]

In [281]:
time_df["54511"] = [len(ndf.loc[(ndf["일시"] == i) & (ndf["지점"] == 54511)]) for i in time_df["datetime"]]
time_df["54823"] = [len(ndf.loc[(ndf["일시"] == i) & (ndf["지점"] == 54823)]) for i in time_df["datetime"]]
time_df["54857"] = [len(ndf.loc[(ndf["일시"] == i) & (ndf["지점"] == 54857)]) for i in time_df["datetime"]]
time_df["58027"] = [len(ndf.loc[(ndf["일시"] == i) & (ndf["지점"] == 58027)]) for i in time_df["datetime"]]
time_df["58238"] = [len(ndf.loc[(ndf["일시"] == i) & (ndf["지점"] == 58238)]) for i in time_df["datetime"]]
time_df["58362"] = [len(ndf.loc[(ndf["일시"] == i) & (ndf["지점"] == 58362)]) for i in time_df["datetime"]]
time_df["58666"] = [len(ndf.loc[(ndf["일시"] == i) & (ndf["지점"] == 58666)]) for i in time_df["datetime"]]
time_df["58847"] = [len(ndf.loc[(ndf["일시"] == i) & (ndf["지점"] == 58847)]) for i in time_df["datetime"]]
time_df["59287"] = [len(ndf.loc[(ndf["일시"] == i) & (ndf["지점"] == 59287)]) for i in time_df["datetime"]]

In [290]:
time_df.head()

datetime  count  54511  54823  54857  58027  58238  58362  58666  \
0 2023-01-01 00:00:00      9      1      1      1      1      1      1      1   
1 2023-01-01 03:00:00      9      1      1      1      1      1      1      1   
2 2023-01-01 06:00:00      9      1      1      1      1      1      1      1   
3 2023-01-01 09:00:00      9      1      1      1      1      1      1      1   
4 2023-01-01 12:00:00      9      1      1      1      1      1      1      1   

   58847  59287  
0      1      1  
1      1      1  
2      1      1  
3      1      1  
4      1      1

In [293]:
time_df.columns[2:]

Index(['54511', '54823', '54857', '58027', '58238', '58362', '58666', '58847',
       '59287'],
      dtype='object')

In [294]:
time_df["datetime"]

0      2023-01-01 00:00:00
1      2023-01-01 03:00:00
2      2023-01-01 06:00:00
3      2023-01-01 09:00:00
4      2023-01-01 12:00:00
               ...        
2915   2023-12-31 09:00:00
2916   2023-12-31 12:00:00
2917   2023-12-31 15:00:00
2918   2023-12-31 18:00:00
2919   2023-12-31 21:00:00
Name: datetime, Length: 2920, dtype: datetime64[ns]

In [305]:
missing_data = []

for i in time_df.columns[2:]:
    for j in time_df["datetime"]:
        if time_df.set_index("datetime").loc[j, str(i)] == 0:
            missing_data.append([j, i])
            
print(len(missing_data))

3920


In [306]:
sandwich_data = []

for i in time_df.columns[2:]:
    for j in time_df["datetime"]:
        if time_df.set_index("datetime").loc[j, str(i)] == 0 and time_df.set_index("datetime").loc[j - timedelta(hours = 3), str(i)] == 1 and time_df.set_index("datetime").loc[j + timedelta(hours = 3), str(i)] == 1:
            sandwich_data.append([j, i])
            
print(len(sandwich_data))

386


In [316]:
str(sandwich_data[0][0].hour).zfill(2)

'03'

In [308]:
ndf2 = ndf.copy()

In [309]:
ndf2.head()

지점                  일시  16 방위 풍향   풍속  강수량   기온 month time
0  54511 2023-01-01 00:00:00       6.0  1.0  0.0 -5.9    01   00
1  54511 2023-01-01 03:00:00       4.0  3.0  0.0  1.2    01   03
2  54511 2023-01-01 06:00:00       6.0  3.0  0.0  3.4    01   06
3  54511 2023-01-01 09:00:00      22.0  1.0  0.0  1.2    01   09
4  54511 2023-01-01 12:00:00      36.0  0.0  0.0 -4.3    01   12

In [339]:
ndf2.dtypes

지점                   int64
일시          datetime64[ns]
16 방위 풍향           float64
풍속                 float64
강수량                float64
기온                 float64
month               object
time                object
dtype: object

In [320]:
ndf3 = []

for date, loc in sandwich_data:
    ndf3.append([loc, date, np.NaN, np.NaN, np.NaN, np.NaN, str(date.month).zfill(2), str(date.hour).zfill(2)])

In [326]:
ndf3 = pd.DataFrame(ndf3)

In [327]:
ndf3.columns = ndf2.columns

In [342]:
ndf3["지점"] = ndf3["지점"].astype(int)

In [343]:
ndf3.head()

지점                  일시  16 방위 풍향  풍속  강수량  기온 month time
0  54511 2023-01-05 03:00:00       NaN NaN  NaN NaN    01   03
1  54511 2023-01-18 06:00:00       NaN NaN  NaN NaN    01   06
2  54511 2023-01-20 15:00:00       NaN NaN  NaN NaN    01   15
3  54511 2023-02-11 18:00:00       NaN NaN  NaN NaN    02   18
4  54511 2023-04-01 12:00:00       NaN NaN  NaN NaN    04   12

In [401]:
ndf4 = pd.concat([ndf2, ndf3], axis = 0).sort_values(by = ["지점", "일시"]).reset_index(drop = True)

In [430]:
def average_angle_ignore_nan(degrees):
    """
    주어진 각도의 리스트에서 NaN 값을 무시하고 평균을 계산합니다.
    
    :param degrees: 각도의 리스트 (0-360도)
    :return: 평균 각도 (0-360도)
    """
    # NaN 값을 무시하고 유효한 각도만 선택
    valid_degrees = [deg for deg in degrees if not pd.isna(deg)]
    
    if not valid_degrees:
        return np.nan  # 유효한 각도가 없는 경우 NaN 반환
    
    # 유효한 각도를 라디안으로 변환
    radians = np.deg2rad(valid_degrees)
    
    # x, y 좌표 계산
    x_coords = np.cos(radians)
    y_coords = np.sin(radians)
    
    # x, y 좌표의 평균 계산
    x_mean = np.mean(x_coords)
    y_mean = np.mean(y_coords)
    
    # 평균 좌표를 각도로 변환
    mean_rad = np.arctan2(y_mean, x_mean)
    mean_deg = np.rad2deg(mean_rad)
    
    # 결과를 0-360도 사이의 값으로 변환
    mean_deg = mean_deg % 360
    
    return mean_deg

In [402]:
ndf4.head()

지점                  일시  16 방위 풍향   풍속  강수량   기온 month time
0  54511 2023-01-01 00:00:00       6.0  1.0  0.0 -5.9    01   00
1  54511 2023-01-01 03:00:00       4.0  3.0  0.0  1.2    01   03
2  54511 2023-01-01 06:00:00       6.0  3.0  0.0  3.4    01   06
3  54511 2023-01-01 09:00:00      22.0  1.0  0.0  1.2    01   09
4  54511 2023-01-01 12:00:00      36.0  0.0  0.0 -4.3    01   12

In [403]:
# 풍속 평균 처리
for i in range(len(ndf4)):
    if pd.isna(ndf4.loc[i, "풍속"]):
        start = max(0, i-2)
        end = min(len(ndf4), i+2)
        arr = list(range(start, end + 1))
        arr.remove(i)
        ndf4.loc[i, "풍속"] = round(ndf4.loc[arr, "풍속"].mean(), 0)

In [404]:
for i in range(len(ndf4)):
    if pd.isna(ndf4.loc[i, "강수량"]):
        start = max(0, i-2)
        end = min(len(ndf4), i+2)
        arr = list(range(start, end + 1))
        arr.remove(i)
        ndf4.loc[i, "강수량"] = round(ndf4.loc[arr, "강수량"].mean(), 1)

In [409]:
for i in range(len(ndf4)):
    if pd.isna(ndf4.loc[i, "기온"]):
        start = max(0, i-2)
        end = min(len(ndf4), i+2)
        arr = list(range(start, end + 1))
        arr.remove(i)
        ndf4.loc[i, "기온"] = round(ndf4.loc[arr, "기온"].mean(), 1)

In [410]:
ndf4[ndf4["16 방위 풍향"].isna()]

지점                  일시  16 방위 풍향   풍속  강수량    기온 month time
33     54511 2023-01-05 03:00:00       NaN  2.0  0.0  -1.7    01   03
138    54511 2023-01-18 06:00:00       NaN  2.0  0.0  -1.8    01   06
157    54511 2023-01-20 15:00:00       NaN  2.0  0.0  -6.8    01   15
334    54511 2023-02-11 18:00:00       NaN  1.0  0.1   2.7    02   18
589    54511 2023-04-01 12:00:00       NaN  2.0  0.0  17.0    04   12
...      ...                 ...       ...  ...  ...   ...   ...  ...
22612  59287 2023-12-15 00:00:00       NaN  1.0  0.0  24.0    12   00
22637  59287 2023-12-18 03:00:00       NaN  2.0  0.0   9.9    12   03
22645  59287 2023-12-19 09:00:00       NaN  3.0  0.0   9.6    12   09
22708  59287 2023-12-27 06:00:00       NaN  2.0  0.0  17.0    12   06
22712  59287 2023-12-27 18:00:00       NaN  2.0  0.0  14.0    12   18

[386 rows x 8 columns]

In [434]:
for i in range(len(ndf4)):
    if pd.isna(ndf4.loc[i, "16 방위 풍향"]):
        start = max(0, i-2)
        end = min(len(ndf4), i+2)
        arr = list(range(start, end + 1))
        arr.remove(i)
        ndf4.loc[i, "16 방위 풍향"] = round(average_angle_ignore_nan(ndf4.loc[arr, "16 방위 풍향"].map(lambda x: x*10).tolist())/10, 0)

In [436]:
time_list2 = [[ndf4.loc[0, "일시"] + timedelta(hours = 3) * i] for i in range(365*8)]

In [437]:
for i in range(len(time_list2)):
    time_list2[i].append(len(ndf4[ndf4["일시"] == time_list2[i][0]]))

In [438]:
time_df2 = pd.DataFrame(time_list2)

In [439]:
time_df2[1].map(lambda x: abs(9 - x)).sum()

3534

In [440]:
time_df2.columns = ["datetime", "count"]

In [441]:
time_df2["54511"] = [len(ndf4.loc[(ndf4["일시"] == i) & (ndf4["지점"] == 54511)]) for i in time_df2["datetime"]]
time_df2["54823"] = [len(ndf4.loc[(ndf4["일시"] == i) & (ndf4["지점"] == 54823)]) for i in time_df2["datetime"]]
time_df2["54857"] = [len(ndf4.loc[(ndf4["일시"] == i) & (ndf4["지점"] == 54857)]) for i in time_df2["datetime"]]
time_df2["58027"] = [len(ndf4.loc[(ndf4["일시"] == i) & (ndf4["지점"] == 58027)]) for i in time_df2["datetime"]]
time_df2["58238"] = [len(ndf4.loc[(ndf4["일시"] == i) & (ndf4["지점"] == 58238)]) for i in time_df2["datetime"]]
time_df2["58362"] = [len(ndf4.loc[(ndf4["일시"] == i) & (ndf4["지점"] == 58362)]) for i in time_df2["datetime"]]
time_df2["58666"] = [len(ndf4.loc[(ndf4["일시"] == i) & (ndf4["지점"] == 58666)]) for i in time_df2["datetime"]]
time_df2["58847"] = [len(ndf4.loc[(ndf4["일시"] == i) & (ndf4["지점"] == 58847)]) for i in time_df2["datetime"]]
time_df2["59287"] = [len(ndf4.loc[(ndf4["일시"] == i) & (ndf4["지점"] == 59287)]) for i in time_df2["datetime"]]

In [445]:
incomplete_time = time_df2[time_df2["count"] != 9]["datetime"].tolist()

In [449]:
ndf4.head()

지점                  일시  16 방위 풍향   풍속  강수량   기온 month time
0  54511 2023-01-01 00:00:00       6.0  1.0  0.0 -5.9    01   00
1  54511 2023-01-01 03:00:00       4.0  3.0  0.0  1.2    01   03
2  54511 2023-01-01 06:00:00       6.0  3.0  0.0  3.4    01   06
3  54511 2023-01-01 09:00:00      22.0  1.0  0.0  1.2    01   09
4  54511 2023-01-01 12:00:00      36.0  0.0  0.0 -4.3    01   12

In [453]:
ndf5 = ndf4[~ndf4["일시"].isin(incomplete_time)]

In [454]:
ndf5.shape

(22392, 8)

In [459]:
(365*8*9 - ndf5.shape[0]) / (365*8*9)

0.14794520547945206

In [460]:
ndf5.to_csv("./중국2023기상데이터.csv")